batch2/batch/driver/scheduler.py

```diff
@@ -19,6 +20,65 @@ def __init__(self, app):
         asyncio.ensure_future(self.loop('cancel_loop', self.cancel_state_changed, self.cancel_1))
         asyncio.ensure_future(self.bump_loop())
 
+    async def compute_fair_share(self):
+        free_cores_mcpu = sum([worker.free_cores_mcpu for worker in self.inst_pool.healthy_instances_by_free_cores])
+
+        user_running_cores_mcpu = {}
+        user_total_cores_mcpu = {}
+        user_allocated_cores = {}
+
+        pending_users_by_running_cores = sortedcontainers.SortedSet(
+            key=lambda user: user_running_cores_mcpu[user])
+        allocating_users_by_total_cores = sortedcontainers.SortedSet(
+            key=lambda user: user_total_cores_mcpu[user])
+
+        records = self.db.execute_and_fetchall(
+            '''
+SELECT user, ready_cores_mcpu, running_cores_mcpu
+FROM user_resources
+WHERE ready_cores_mcpu > 0;
+''')
+
+        async for record in records:
+            user = record['user']
+            user_running_cores_mcpu[user] = record['running_cores_mcpu']
+            user_total_cores_mcpu[user] = record['running_cores_mcpu'] + record['ready_cores_mcpu']
+            pending_users_by_running_cores.add(user)
+
+        mark = 0
+        while pending_users_by_running_cores:
```

Comment:
This seems wrong.  If there are no pending users but still free cores and allocating users, you should keep going.

Basically, you should stop if: there are no more free cores, or there are no more running or pending users.

Reply:
I fixed this, but I think the way I wrote the code with break statements, it could just be True.